src/number_theory/arithmetic_function.lean

```diff
@@ -153,23 +159,65 @@ instance [add_comm_group R] : add_comm_group (arithmetic_function R) :=
 { .. arithmetic_function.add_comm_monoid,
   .. arithmetic_function.add_group }
 
-section dirichlet_ring
-variable [semiring R]
+section has_scalar
+variables {M : Type*} [has_zero R] [add_comm_monoid M] [has_scalar R M]
+
+/-- The Dirichlet convolution of two arithmetic functions `f` and `g` is another arithmetic function
+  such that `(f * g) n` is the sum of `f x * g y` over all `(x,y)` such that `x * y = n`. -/
+instance : has_scalar (arithmetic_function R) (arithmetic_function M) :=
+⟨λ f g, ⟨λ n, ∑ x in divisors_antidiagonal n, f x.fst • g x.snd, by simp⟩⟩
+
+@[simp]
+lemma smul_apply {f : arithmetic_function R} {g : arithmetic_function M} {n : ℕ} :
+  (f • g) n = ∑ x in divisors_antidiagonal n, f x.fst • g x.snd := rfl
+
+end has_scalar
 
 /-- The Dirichlet convolution of two arithmetic functions `f` and `g` is another arithmetic function
   such that `(f * g) n` is the sum of `f x * g y` over all `(x,y)` such that `x * y = n`. -/
-instance : has_mul (arithmetic_function R) :=
-⟨λ f g, ⟨λ n, ∑ x in divisors_antidiagonal n, f x.fst * g x.snd, by simp⟩⟩
+instance [semiring R] : has_mul (arithmetic_function R) := ⟨(•)⟩
 
 @[simp]
-lemma mul_apply {f g : arithmetic_function R} {n : ℕ} :
+lemma mul_apply [semiring R] {f g : arithmetic_function R} {n : ℕ} :
   (f * g) n = ∑ x in divisors_antidiagonal n, f x.fst * g x.snd := rfl
 
-instance : monoid (arithmetic_function R) :=
-{ one_mul := λ f,
-  begin
+section module
+variables {M : Type*} [semiring R] [add_comm_monoid M] [semimodule R M]
+
+lemma mul_smul' (f g : arithmetic_function R) (h : arithmetic_function M) :
+  (f * g) • h = f • g • h :=
+begin
+    ext n,
+    simp only [mul_apply, smul_apply],
+    have := @finset.sum_sigma (ℕ × ℕ) M _ _ (divisors_antidiagonal n)
+      (λ p, (divisors_antidiagonal p.1)) (λ x, f x.2.1 • g x.2.2 • h x.1.2),
+    convert this.symm using 1; clear this,
+    { apply finset.sum_congr rfl,
+      intros p hp,
+      simp only [sum_smul, mul_smul], },
+    have := @finset.sum_sigma (ℕ × ℕ) M _ _ (divisors_antidiagonal n)
+      (λ p, (divisors_antidiagonal p.2)) (λ x, f x.1.1 • (g x.2.1 • h x.2.2)),
+    convert this.symm using 1; clear this,
```

Comment:
The proof is snappy until this convert, then it starts slowing down.
I haven't profiled in detail, but I expect that the two `finish` calls also cause this to slow down by a lot.
I think we should pay a bit more attention to speed, because
```
time lean src/number_theory/arithmetic_function.lean

real    0m31.573s
user    1m13.497s
sys     0m2.660s
```

Reply:
I've removed both `convert`s and both `finish`s. Hopefully it's faster now.

Reply:
It certainly is faster now. I squeezed some more simp calls, an now I get
```
time lean src/number_theory/arithmetic_function.lean

real    0m19.147s
user    0m51.370s
sys     0m2.035s
```
I don't think this PR is the place for a general speedup. But it would be good to have another go at this at some point.